presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/AddExchangesBelowPartialAggregationOverGroupIdRuleSet.java

```diff
@@ -0,0 +1,349 @@
+import io.prestosql.sql.planner.optimizations.StreamPreferredProperties;
+import io.prestosql.sql.planner.optimizations.StreamPropertyDerivations.StreamProperties;
+import io.prestosql.sql.planner.plan.AggregationNode;
+import io.prestosql.sql.planner.plan.ExchangeNode;
+import io.prestosql.sql.planner.plan.GroupIdNode;
+import io.prestosql.sql.planner.plan.PlanNode;
+import io.prestosql.sql.planner.plan.ProjectNode;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Verify.verify;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableMultiset.toImmutableMultiset;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static io.prestosql.SystemSessionProperties.getTaskConcurrency;
+import static io.prestosql.SystemSessionProperties.isEnableForcedExchangeBelowGroupId;
+import static io.prestosql.SystemSessionProperties.isEnableStatsCalculator;
+import static io.prestosql.matching.Capture.newCapture;
+import static io.prestosql.matching.Pattern.nonEmpty;
+import static io.prestosql.matching.Pattern.typeOf;
+import static io.prestosql.sql.planner.SystemPartitioningHandle.FIXED_HASH_DISTRIBUTION;
+import static io.prestosql.sql.planner.optimizations.StreamPreferredProperties.fixedParallelism;
+import static io.prestosql.sql.planner.optimizations.StreamPropertyDerivations.deriveProperties;
+import static io.prestosql.sql.planner.plan.ExchangeNode.Scope.LOCAL;
+import static io.prestosql.sql.planner.plan.ExchangeNode.Scope.REMOTE;
+import static io.prestosql.sql.planner.plan.ExchangeNode.partitionedExchange;
+import static io.prestosql.sql.planner.plan.Patterns.Aggregation.groupingColumns;
+import static io.prestosql.sql.planner.plan.Patterns.Aggregation.step;
+import static io.prestosql.sql.planner.plan.Patterns.Exchange.scope;
+import static io.prestosql.sql.planner.plan.Patterns.source;
+import static java.lang.Double.isNaN;
+import static java.lang.Math.min;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Transforms
+ * <pre>
+ *   - Exchange
+ *     - [ Projection ]
+ *       - Partial Aggregation
+ *         - GroupId
+ * </pre>
+ * to
+ * <pre>
+ *   - Exchange
+ *     - [ Projection ]
+ *       - Partial Aggregation
+ *         - GroupId
+ *           - LocalExchange
+ *             - RemoteExchange
+ * </pre>
+ * <p>
+ * Rationale: GroupId increases number of rows (number of times equal to number of grouping sets) and then
+ * partial aggregation reduces number of rows. However, under certain conditions, exchanging the rows before
+ * GroupId (before multiplication) makes partial aggregation more effective, resulting in less data being
+ * exchanged afterwards.
+ */
+public class AddExchangesBelowPartialAggregationOverGroupIdRuleSet
```

Comment:
It would be nice to have unit test for that

Reply:
That would probably be actually useless -- https://github.com/prestodb/presto/pull/11741#discussion_r228836050